packages/codemirror-blocks/src/copypaste.ts

```diff
@@ -68,12 +65,11 @@ export function copy(
     text = text + prefix + node.toString();
     postfix = node.options && node.options.comment ? "\n" : " ";
   }
+  // do the actual copy, then restore focus
   copyToClipboard(text);
-  // Copy steals focus. Force it back to the node's DOM element
-  // without announcing via activateByNid().
-  // TODO(pcardune): it would be better to focus on whatever the
-  // previously focused element was, not or not.
-  if (focusedNode) {
-    focusedNode.element?.focus();
+  if (previouslyFocusedElement) {
+    (previouslyFocusedElement as HTMLElement).focus();
+  } else {
+    throw "copy was initiated, but no node was focused";
   }
```

Comment:
Is it actually a problem to call `copy()` when nothing has focus?

Reply:
The `copy()` action can only  be initiated from two files: (1) keymap.tsx and (2) PrimitiveList.tsx. In both files, the action is only called when a node is selected/active. If it gets called with no nodes being passed, that means something has gone wrong inside the library itself.

Reply:
Perhaps, but the thing that has gone wrong has nothing to do with the copy() function. To the extent that this check needs to be done somewhere, it should be done somewhere else... close to where the problem might actually be, maybe in those other files. Also, if this check needs to be done, then it could be done more completely. As written, so long as *any* element has focus when copy is called, then the check will pass.

I'm just thinking about this from an API design perspective. If the function was called `copyCurrentlyFocusedNode()`, then it would make more sense to do this check. But then it wouldn't make sense for the nodes being copied to be passed as parameters since they would presumably be looked up from `document.activeElement`, and it would throw if there was no currently focused node.

If I were to write an example of how to use the `copy()` function, it would make sense for the example to look like this:

```ts
copy([Literal(...), FunctionApp(...)]);
```

But actually running that example would throw. An actually correct example would be:

```ts
document.body.focus();
copy([Literal(...), FunctionApp(...)]);
```

which is confusing. It is non-obvious why the second example works but the first doesn't. The documentation for the function doesn't explain why, and neither does the error I get back. Even reading the code for `copy()` doesn't give me any insights. Even reading the code at the existing call sights for `copy()` doesn't give me any insights. By throwing this error, the function becomes more tightly coupled to the way the rest of the application is written at this particular moment in time, making it harder to test, harder to refactor, and harder to understand.